### PR TITLE
[TOPI] Improve CUDA softmax scheduling

### PR DESCRIPTION
**Pull request by wpan11nv**
- Do not use multiple kernels

- Schedule with warp reductions

- Fixed a bug on the lower warp memory pass

- Fixed warp shuffle intrinsics for the nvptx backend.

Signed-off-by: Wei Pan <weip@nvidia.com>

Thanks for contributing to TVM!   Please refer to guideline https://tvm.apache.org/docs/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/apache/incubator-tvm/blob/master/CONTRIBUTORS.md#reviewers) by @ them in the pull request thread.

### CODE REVIEW

**Comment by wpan11nv** [comment]
Please help to review this PR, @tqchen  and @roastduck

File: src/tir/transforms/lower_warp_memory.cc [hunk 1]
```diff
@@ -213,9 +213,13 @@ class WarpAccessRewriter : protected StmtExprMutator {
...
-    CHECK_EQ(alloc_size % (width_ * warp_coeff_), 0)
-        << "Warp memory must be multiple of the extent of threadIdx.x";
-    warp_group_ = alloc_size / (width_ * warp_coeff_);
+
+    // Align the local memory size. The number of elements may not
+    // be a multiple of width_ * warp_coeff_; round it up.
+    int factor = width_ * warp_coeff_;
+    warp_group_ = (alloc_size + (factor - 1)) / factor;
+    alloc_size = warp_group_ * factor;
+
```
Comment (roastduck) [inline src/tir/transforms/lower_warp_memory.cc:222]: Can you add a test for the non-dividable case in `tests/python/unittest/test_tir_transform_lower_warp_memory.py`? There might be some boundary checking issues.

Comment (wpan11nv) [reply src/tir/transforms/lower_warp_memory.cc:222]: Added tests. Without this fix, this test will assert.

Comment (roastduck) [inline src/tir/transforms/lower_warp_memory.cc:222]: Please consider this scenario:

```
// bound to threadIdx.y
for (i = 0; i < 2; i++) {
  // bound to threadIdx.x
  for (j = 0; j < 5; j++) {
    access buffer.warp[(j + 1) % 5]
  }
}
```

Now we are shuffling among Thread 0\~4, and Thread 5\~9. Since we are rounding them up to Thread 0\~7 and Thread **5\~12 (seems not right as well)**, will there be any conflict between these two group of threads?

Comment (wpan11nv) [reply src/tir/transforms/lower_warp_memory.cc:222]: What’s the extent of threadidx.x? I think the test case in this PR clearly shows rounding up to the warp size is needed. It is the same as softmax failures I have seen. As there is no warp level allocation of size n,  we allocate n/32 elements in each thread. If n is not a multiple of 32, we need to ”over-allocate“ slightly and predict the access.

Comment (roastduck) [reply src/tir/transforms/lower_warp_memory.cc:222]: > What’s the extent of threadidx.x?

It's 5. Thread 0\~4 are shuffled as a group (say Group 0), and Thread 5\~12 are another group (say Group 1). If we round up 0\~4 to 0\~7, Thread 5\~7 may be both accessed by Group 0 and Group 1, which may cause some problems.

> I think the test case in this PR clearly shows rounding up to the warp size is needed.

Yes, it is needed, but maybe not enough.

> As there is no warp level allocation of size n, we allocate n/32 elements in each thread. If n is not a multiple of 32, we need to ”over-allocate“ slightly and predict the access.

Yes, we need to over-allocate the **buffers**, but should we also over-allocate the **threads**? If not, maybe we should properly set the mask of `__shlf_sync`.

Comment (wpan11nv) [reply src/tir/transforms/lower_warp_memory.cc:222]: Looks like this is a sub-warp related, and I am not following it closely. CUDA __shfl_sync does not support non-power of 2 width. So your example may not be supported.  It would be helpful if we have a legal example that would be broken by this patch. Thanks!

Comment (roastduck) [reply src/tir/transforms/lower_warp_memory.cc:222]: Sorry I misunderstood these lines. I thought it was dealing with sub-warp logic. Then there will be no problem.

Comment (wpan11nv) [reply src/tir/transforms/lower_warp_memory.cc:222]: Yes, it is not sub-warp specific. Thanks!

File: src/target/llvm/codegen_llvm.cc [hunk 2]
```diff
@@ -736,7 +736,40 @@ llvm::Function* CodeGenLLVM::GetIntrinsicDecl(llvm::Intrinsic::ID id, llvm::Type
 #endif  // TVM_LLVM_VERSION
 }
 
+// Check if this is a warp shuffle intrinsic call and match its
+// corresponding nvvm intrinsic. Return false if the match is successful.
+static bool get_warp_shuffle_intrinsic(const CallNode* op, llvm::Intrinsic::ID* id) {
```
Comment (tqchen) [inline src/target/llvm/codegen_llvm.cc:741]: Use CamelCase per Google C style

**Comment by tqchen** [comment]
@icemelon9 @Hzfengsy @Shawn-Inspur  please also take a look

File: topi/python/topi/cuda/softmax.py [hunk 3]
```diff
@@ -53,13 +54,62 @@ def schedule_softmax(outs):
...
+        k = expsum.op.reduce_axis[0]
+        ko, _ = s[expsum].split(k, nparts=num_thread)
+        s[expsum].bind(ko, thread_x)
+        s[expsum].compute_at(s[softmax], xo)
+
+        # (2) exp
+        if exp is not None:
+            xo, xi = s[exp].split(exp.op.axis[1], nparts=num_thread)
+            _, xii = s[exp].split(xi, factor=4)
+            s[exp].vectorize(xii)
```
Comment (Shawn-Inspur) [inline topi/python/topi/cuda/softmax.py:97]: I am wondering whether the vectorizing of “xii” should be included in an if block, which checked "tgt.target_name != "nvptx", for the logic here looks the same as presented in line 81.

File: topi/python/topi/cuda/softmax.py [hunk 4]
```diff
@@ -53,13 +54,61 @@ def schedule_softmax(outs):
         raise ValueError('Tag is expected to be softmax_output or log_softmax_output. \
                          Got {0}'.format(op_tag))
 
+    # The nvptx backend only supports 32-bits warp shuffle instructions.
+    #
+    # TODO(tvm-team) Fix nvptx codegen or deprecate nvptx backend.
+    def sched_warp_softmax():
+        if tgt.target_name == "nvptx":
+            return softmax.dtype == "float32" or softmax.dtype == "int32"
+        return True
```
Comment (Shawn-Inspur) [inline topi/python/topi/cuda/softmax.py:63]: This makes sense to me.

**Comment by tqchen** [comment]
@Shawn-Inspur please followup and https://tvm.apache.org/docs/contribute/code_review.html#approve-and-request-changes-explicitly

**Comment by wpan11nv** [comment]
This looks like a doc error. Let me trigger the build again.

File: topi/python/topi/cuda/softmax.py (hunk 3 above)
Comment (wpan11nv) [reply topi/python/topi/cuda/softmax.py:96]: Good point, I forgot why I added this nvptx check. Now removed,

**Comment by wpan11nv** [comment]
This build is fine now and all comments are addressed. Thanks all for reviewing this patch!

**Comment by tqchen** [comment]
THanks @wpan11nv , @roastduck @Shawn-Inspur !

**Comment by t-vi** [comment]
This broke the ROCm backend.

**Comment by tqchen** [comment]
@t-vi can you elaborate?

@wpan11nv please also followup. One thing that get missed during review is that we should move the nvptx related intrinsics into codegen_nvptx.cc instead.

File: src/target/llvm/codegen_llvm.cc [hunk 5]
```diff
@@ -736,7 +736,40 @@ llvm::Function* CodeGenLLVM::GetIntrinsicDecl(llvm::Intrinsic::ID id, llvm::Type
 #endif  // TVM_LLVM_VERSION
 }
 
+// Check if this is a warp shuffle intrinsic call and match its
+// corresponding nvvm intrinsic. Return true if the match is successful.
+static bool GetWarpShuffleIntrinsic(const CallNode* op, llvm::Intrinsic::ID* id) {
+  // Only 32 bit data type is supported.
+  if (op->dtype.is_vector() || op->dtype.bits() != 32) {
```
Comment (tqchen) [inline src/target/llvm/codegen_llvm.cc:743]: We should move this to the codegen_nvptx.cc instead and overload the intrinsics creation.

File: topi/python/topi/cuda/softmax.py [hunk 6]
```diff
@@ -39,6 +39,7 @@ def schedule_softmax(outs):
     outs = [outs] if isinstance(outs, te.tensor.Tensor) else outs
     s = te.create_schedule([x.op for x in outs])
     softmax = outs[0]
+    tgt = target_.Target.current(allow_none=False)
```
Comment (tqchen) [inline topi/python/topi/cuda/softmax.py:42]: We might want to register the warp level strategies only when the target is cuda

**Comment by t-vi** [comment]
So ROCm uses the CUDA schedule, but warp reductions don't seem to currently work (so arguably, ROCm would want to be improved). But so before this PR, one could run resnet18 with rocm backend and now one cannot.
This can also be seen earlier, when running the warp reduction tests on ROCm.
I've looked a bit into fixing it, but I haven't fully understood from which of the three related patches this stems. 
(Incidentally, it triggered also a corner case for me on cuda where nvrtc would accidentally use an cuda-8.0 instead of the 10.1 the libnvrtc belonged to.)

**Comment by wpan11nv** [comment]
> So ROCm uses the CUDA schedule, but warp reductions don't seem to currently work (so arguably, ROCm would want to be improved). But so before this PR, one could run resnet18 with rocm backend and now one cannot.
> This might alsscheduleo be seen earlier, when running the tests on ROCm.
> I've looked a bit into fixing it, but I haven't fully understood from which of the three related patches this stems.
> (Incidentally, it triggered also a corner case for me on cuda where nvrtc would accidentally use an cuda-8.0 instead of the 10.1 the libnvrtc belonged to.)

Can we disable this ROCm here?
https://github.com/apache/incubator-tvm/blob/master/topi/python/topi/cuda/softmax.py#L60

File: src/target/llvm/codegen_llvm.cc (hunk 5 above)
Comment (wpan11nv) [reply src/target/llvm/codegen_llvm.cc:743]: Sure. I will have a look.

**Comment by t-vi** [comment]
I'm adding shfl intrinsics to the rocm bits (using `tvm.intrin.rule.rocm.tvm_warp_shuffle /-up/-down` definitions).
I'm currently seeing a funny effect where I get a `tvm_thread_allreduce` call with null arguments in `lower_thread_allreduce`'s `MakeAllreduce`. Eventually, I hope to get to the codegen - when I'll probably run into the nvptx bits in the llvm codegen. Is there a reason not to use the intrin.rule mechanism for nvptx?
I'm not sure running `gpu_imagenet_bench.py` (which I'm using as the first stop of seeing if anything works) with the nvptx target works for me (though I get to the codegen for that), but I would not know if it worked before...

File: src/target/llvm/codegen_llvm.cc (hunk 5 above)
Comment (t-vi) [reply src/target/llvm/codegen_llvm.cc:743]: I'll can do this move when I fix rocm. (I needed to do it anyways, so...)

File: topi/python/topi/cuda/softmax.py (hunk 6 above)
Comment (t-vi) [reply topi/python/topi/cuda/softmax.py:42]: Hm. I'd like to benefit from the GPU schedules with warps...

**Comment by t-vi** [comment]
@wpan11nv  Thanks for your offer to help. I submitted the clean-up #5726 and then in #5727 I add ROCm warp reductions. One of the things I did was to avoid assuming a fixed warp-size of 32 in the TIR transformations before codegen.
Thank you for improving softmax btw - it was something that looked funny with the four kernels before.

**Comment by tqchen** [comment]
cc @icemelon9 it might be useful to revisit the softmax strategy.
